rfcs/rfc0020.md

```diff
@@ -0,0 +1,116 @@
+
+A new feature of the Perl 5.18 release was the ability to create lexically-named subroutines, whose names are accessible within some lexical scope but not visible in the package namespace, and thus not visible to callers from outside the code, even via method reflection. Since then, the ability has not been further expanded on by Perl core. While a few CPAN modules have tried to implement the ability, the idea does not seem to have caught on in the majority of cases.
+
+The `builtin` module added in Perl 5.36 was the first core module to export its named functions lexically into the calling scope, rather than symbolically into the caller's package. It aims to set a new standard for better-behaved module exporting, but is currently handled by its own internal implementation which is not yet exposed in a way that other modules can make use of.
+
+## Rationale
+
+By providing a mechanism by which other modules written in pure perl code can take advantage of lexical export themselves, we hope that more will begin to take advantage of this ability. This will lead to modules that behave more politely in the face of the growing collection of modules written in object-oriented style. In addition, as addressed by the "Future Scope" section, we hope this leads on to the creation of better Exporter-like modules or other core perl abilities to take advantage of lexical export.
+
+## Specification
+
+A new function in the `builtin` package, called `lexically_export`. This function takes an even-sized name/value list of pairs. It does not return anything.
+
+```perl
+builtin::lexically_export $name1, $value1, $name2, $value2, ...;
+```
+
+Each name/value pair acts independently. The name gives the new name to be created in the scope being compiled, and the value is a reference to the item the name should have.
+
+Four kinds of items are supported:
+
++ If the value is a `CODE` reference, a new named subroutine will be created in the scope. The corresponding name argument may optionally begin with a `&` sigil.
+
++ If the name is a `SCALAR` reference, a new scalar variable will be created in the scope. The corresponding name argument must begin with a `$` sigil.
+
++ If the name is an `ARRAY` reference, a new array variable will be created in the scope. The corresponding name argument must begin with a `@` sigil.
+
++ If the name is a `HASH` reference, a new hash variable will be created in the scope. The corresponding name argument must begin with a `%` sigil.
+
+If no sigil is present on the name argument, it is presumed to be of the first kind, and the value must be a `CODE` reference.
+
+As the only effect this function has is to add new things to the compiletime lexical scope, it shall be an error to call this function during regular runtime, when no code is currently being compiled. It shall also be an error to pass non-reference values, references to unsupported items (e.g. globs), or references whose type does not match the sigil of its name.
```

Comment:
What would the alternative be?

Reply:
It has to be *a* compiletime scope, because that's the only thing that can work. It's unlikely there are multiple nested ones, and in any case I think it would be weird if we were to try to allow some kind of "uplevel" ability to affect scopes beyond the immediate caller. I don't think there's any other way we could offer.